### DISCUSSION TITLE
Developers: Continue Supporting Python 3.7?

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvistaqt#186] [Q&A]
(@pyvista/developers) @tkoyama010 @akaszynski @adeak 

Do we want to continue supporting Python 3.7 for `pyvistaqt`? There are several packages that won't get updated with 3.7, namely:

- ipython: 7.34.0 -> 8.4.0
- numpy: 1.21.6 -> 1.23.0
- pylint: 2.13.0 -> 2.14.3 

It looks like support will end in June of next year, so we have time: [PEP 537](https://peps.python.org/pep-0537/#lifespan). However, it would be nice to bump up.

**Comment by adeak** [comment]
What specific features are you aiming for with 3.8?

**Reply by adam-grant-hendry** [reply]
- `IPython.kernel` was deprecated in IPython 8.0. This was a wrapper for `ipykernel`, so when Python 3.7 sunsets in  a year, `pyvistaqt` will need updating to accomodate the new major changes. (furthermore, `ipykernel` is type hinted, whereas `IPython` is not)

- I've helped fix a type hinting bug for Qt slots/signals in `pylint` via `astroid`, which would be helpful to have (see [PR #1654](https://github.com/PyCQA/astroid/pull/1654); probably will make its way in in 2.14.4 though)

Namely, these will help me update type hinting. I can try and get around it for now if we really want to still keep 3.7, but I'll have to fix certain things because dependabot can't bump things up:

- [Bump pylint from 2.13.0 to 2.14.3 #175](https://github.com/pyvista/pyvistaqt/pull/175)
- [Bump numpy from 1.21.6 to 1.23.0 #176](https://github.com/pyvista/pyvistaqt/pull/176)
- [Bump ipython from 7.34.0 to 8.4.0 #180](https://github.com/pyvista/pyvistaqt/pull/180)

I want to see if I should go through the effort if 3.7 is only going to be around 1 more year.

**Reply by adeak** [reply]
Doesn't the first one mean that we have to look for a replacement anyway? In other words, can we replace `IPython.kernel` irrespective of a Python version requirement bump?

**Reply by adam-grant-hendry** [reply]
Only the `kernel` module in `IPython` was deprecated. The first one can be handled very simply by using the `ipykernel` package everywhere `IPython.kernel` pops up in the source and adding it to `requirements_test.txt`. Python 3.7 supports the latest version of `ipykernel`. For that, I'm not terribly concerned. (We should start supporting that transition now though)

For the other, everywhere signals and slots occur and the backend used is PySide6, `pylint` is going to spit out an error, which might keep me from pushing up changes now that we implemented precommit for `pyvistaqt`

**Reply by adam-grant-hendry** [reply]
As an aside, in `ci/azure-pipelines.yml`, I see we only test against `PyQt5==5.11.3`. (I know from the docs that we've historically used `PyQt5` and are transitioning to `qtpy`; the `PyQt` licenses have historically been restrictive, which has prevented us in the past from adding them to the `requirements_test.txt`.)

In transitioning to `qtpy`, there are a few subtle testing ramifications:

- There are several differences between PyQt5 < 5.11 and > 5.11 that cause `mypy` to fail in > 5.11, and `PySide6` has fixed several type hinting errors and bugs that plagued `pytestqt` (see my answer to [`pytest-qt` Function `mouseMove()` Not Working](https://stackoverflow.com/questions/68696865/pytest-qt-function-mousemove-not-working/72536434#72536434) ).

- For the unsuspecting tester, there is a subtle bug if you install all Qt binding versions and attempt to test (see [BUG: Multiple PyQt/PySide Versions Not Supported #437](https://github.com/pytest-dev/pytest-qt/issues/437#issue-1286042797) ). Tests simply won't run and `pytest` fails silenty.

**Comment by tkoyama010** [comment]
Thank you for bringing this to our attention. My personal opinion is that PyVista and PyVistaQt should keep pace with the [Python end of life](https://endoflife.date/python). This means that Python 3.7 support should continue until 27 Jun 2023. If there is a problem with package dependencies, I think we should stop updating.

**Reply by adam-grant-hendry** [reply]
@tkoyama010 Okay, thanks for confirming! No concerns here. May I submit a PR to change to using `ipykernel` and workarounds for `pylint` that will still work for Python 3.7?

**Reply by tkoyama010** [reply]
Pull Request is always welcome.
